packages/core/src/__tests__/acceptance/service.acceptance.ts

```diff
@@ -0,0 +1,143 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/core
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Context} from '@loopback/context';
+import {expect} from '@loopback/testlab';
+import {service} from '../..';
+
+describe('@service', () => {
+  let ctx: Context;
+
+  beforeEach(givenContext);
+
+  it('injects a service using constructor with serviceClass argument', async () => {
+    class MyController {
+      constructor(@service(MyService) public myService: MyService) {}
+    }
+    ctx.bind('controllers.MyController').toClass(MyController);
+
+    const controller = await ctx.get<MyController>('controllers.MyController');
+    expect(controller.myService).to.be.instanceOf(MyService);
+  });
+
+  it('injects a service using property', async () => {
+    class MyController {
+      @service(MyService) public myService: MyService;
+    }
+    ctx.bind('controllers.MyController').toClass(MyController);
+
+    const controller = await ctx.get<MyController>('controllers.MyController');
+    expect(controller.myService).to.be.instanceOf(MyService);
+  });
+
+  it('injects a service without serviceClass argument', async () => {
+    class MyController {
+      constructor(@service() public myService: MyService) {}
+    }
+    ctx.bind('controllers.MyController').toClass(MyController);
+
+    const controller = await ctx.get<MyController>('controllers.MyController');
+    expect(controller.myService).to.be.instanceOf(MyService);
+  });
+
+  it('injects a service matching a sub class', async () => {
+    class MyController {
+      constructor(@service(MyService) public myService: MyService) {}
+    }
+    ctx.unbind('services.MyService');
+    ctx.bind('services.MySubService').toClass(MySubService);
+    ctx.bind('controllers.MyController').toClass(MyController);
```

Comment:
I am concerned about the implications of this feature on user (developer) experience. Personally, I would find it surprising that `@service(MyService)` is resolved using the binding at key `services.MySubService`. If you look at the code wiring up the dependencies, there is no indication whatsoever that binding `services.MySubService` will automatically setup resolution of `services.MyService` too.

```ts
ctx.bind('services.MySubService').toClass(MySubService);
```

Let's find a better solution please. For example, we can make it explicit that one service class will be injected in places asking for another class.

```ts
// Resolve `MyService` service using `MySubService` implementation
ctx.service(MyService).toAlias(MySubService);

// Or maybe the other way around?
ctx.service(MySubService, {aliases: [MyService]})
```

Alternatively, let's keep handling of subclasses out of scope of this initial pull implementation and discuss it in a follow-up pull request.